packages/ckeditor5-ui/src/focuscycler.ts

```diff
@@ -228,7 +231,13 @@ export default class FocusCycler {
 	 * **Note**: Hidden views (e.g. with `display: none`) are ignored.
 	 */
 	public focusNext(): void {
-		this._focus( this.next );
+		const next = this.next;
+
+		this._focus( next, 1 );
+
+		if ( next == this.first ) {
+			this.fire<FocusCyclerForwardCycleEvent>( 'forwardCycle' );
```

Comment:
I am no expert in the UI space, so maybe you'll be able to explain it all to me, but...

AFAICS, the way it works is that after reaching the last focusable element and firing `focusNext()` I will first focus the first element and then fire `forwardCycle` (and possibly immediately focus something else).

This doesn't seem correct, it seems unexpected and I am afraid of side-effects (or hard to discover bugs).

Instead, I'd assume that if we reach the end of focus cycle, `forwardCycle` is fired (instead of straight focusing the view). Then, there's a default listener on low priority that focuses the first element. Finally, to chain focus cyclers, you would overwrite `forwardCycle` and stop the event. We could have a helper method for that, or have it somehow automated, if possible. Same for `backwardCycle`.

This would be also more in-line with how we do things in CKE5.

Alternatively, we could have something like this:

```javascript
this.fire( 'forwardCycle' );

if ( !somethingHappenedOnForwardCycle ) {
    this._focus( first, 1 );
}
```

That is, instead of having a default action on low priority listener, we could focus the first element if `forwardCycle` has not been handled. In this way, you would not need to stop `forwardCycle` event, which would make the handling a little bit easier. But I am afraid that checking if `fowardCycle` was handled in a custom way may be difficult :thinking:. So maybe that's overengineering. Especially since, as I mentioned, in most places in CKE5 we expect people to stop the event after handling it in a custom way. We could also just put it in the API docs for the event.